mempool/clist_mempool.go

```diff
@@ -0,0 +1,712 @@
+	} else {
+		mempool.cache = nopTxCache{}
+	}
+	proxyAppConn.SetResponseCallback(mempool.globalCb)
+	for _, option := range options {
+		option(mempool)
+	}
+	return mempool
+}
+
+// NOTE: not thread safe - should only be called once, on startup
+func (mem *CListMempool) EnableTxsAvailable() {
+	mem.txsAvailable = make(chan struct{}, 1)
+}
+
+// SetLogger sets the Logger.
+func (mem *CListMempool) SetLogger(l log.Logger) {
+	mem.logger = l
+}
+
+// WithPreCheck sets a filter for the mempool to reject a tx if f(tx) returns
+// false. This is ran before CheckTx.
+func WithPreCheck(f PreCheckFunc) Option {
+	return func(mem *CListMempool) { mem.preCheck = f }
+}
+
+// WithPostCheck sets a filter for the mempool to reject a tx if f(tx) returns
+// false. This is ran after CheckTx.
+func WithPostCheck(f PostCheckFunc) Option {
+	return func(mem *CListMempool) { mem.postCheck = f }
+}
+
+// WithMetrics sets the metrics.
+func WithMetrics(metrics *Metrics) Option {
+	return func(mem *CListMempool) { mem.metrics = metrics }
+}
+
+// InitWAL creates a directory for the WAL file and opens a file itself.
+//
+// *panics* if can't create directory or open file.
+// *not thread safe*
+func (mem *CListMempool) InitWAL() {
+	walDir := mem.config.WalDir()
+	err := cmn.EnsureDir(walDir, 0700)
+	if err != nil {
+		panic(errors.Wrap(err, "Error ensuring WAL dir"))
+	}
+	af, err := auto.OpenAutoFile(walDir + "/wal")
+	if err != nil {
+		panic(errors.Wrap(err, "Error opening WAL file"))
+	}
+	mem.wal = af
+}
+
+// CloseWAL closes and discards the underlying WAL file.
+// Any further writes will not be relayed to disk.
+func (mem *CListMempool) CloseWAL() {
+	mem.proxyMtx.Lock()
+	defer mem.proxyMtx.Unlock()
+
+	if err := mem.wal.Close(); err != nil {
+		mem.logger.Error("Error closing WAL", "err", err)
+	}
+	mem.wal = nil
+}
+
+func (mem *CListMempool) Lock() {
+	mem.proxyMtx.Lock()
+}
+
+func (mem *CListMempool) Unlock() {
+	mem.proxyMtx.Unlock()
+}
+
+func (mem *CListMempool) Size() int {
+	return mem.txs.Len()
+}
+
+func (mem *CListMempool) TxsBytes() int64 {
+	return atomic.LoadInt64(&mem.txsBytes)
+}
+
+func (mem *CListMempool) FlushAppConn() error {
+	return mem.proxyAppConn.FlushSync()
+}
+
+func (mem *CListMempool) Flush() {
+	mem.proxyMtx.Lock()
+	defer mem.proxyMtx.Unlock()
+
+	mem.cache.Reset()
+
+	for e := mem.txs.Front(); e != nil; e = e.Next() {
+		mem.txs.Remove(e)
+		e.DetachPrev()
+	}
+
+	mem.txsMap = sync.Map{}
+	_ = atomic.SwapInt64(&mem.txsBytes, 0)
+}
+
+func (mem *CListMempool) TxsFront() *clist.CElement {
+	return mem.txs.Front()
+}
+
+func (mem *CListMempool) TxsWaitChan() <-chan struct{} {
+	return mem.txs.WaitChan()
+}
+
+// It blocks if we're waiting on Update() or Reap().
+// cb: A callback from the CheckTx command.
+//     It gets called from another goroutine.
+// CONTRACT: Either cb will get called, or err returned.
+func (mem *CListMempool) CheckTx(tx types.Tx, cb func(*abci.Response)) (err error) {
```

Comment:
remove?

Reply:
Probably a good idea to discuss this in a separate issue.

Reply:
https://github.com/tendermint/tendermint/issues/3590